[docs] Update `interactionPolicy` (#3703)
superseriousbusiness/gotosocial: docs/federation/posts.md

## Interaction Policy

GoToSocial uses the property `interactionPolicy` on posts in order to indicate to remote instances what sort of interactions are (conditionally) permitted to be processed and stored by the origin server, for any given post.

The `@context` document for `interactionPolicy` is at `https://gotosocial.org/ns`.

Given that not all ActivityPub server softwares know about / support `interactionPolicy` yet, the presence of the `interactionPolicy` property on a federated post may be considered **advisory** until the extension is more widely adopted. Ie., `interactionPolicy` itself is separate from, and plays no part in, the approval/rejection flow described below

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Author: > rather, we are lacking a mechanism to say that we expect verification.

Yes that's true! I wonder if something could be added somewhere :thinking: Some kind of `mustVerify` or `mustValidate` property or something like that? I'm not sure, just spitballing.

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Reviewer: What I understood and what I would advocate for is the presence of the policy implying that it should be verified with the according mechanism. This makes the most sense to me, but it does require the ability to leave the policy out without that meaning “you can't interact with it”.

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Author: > the presence of the policy implying that it should be verified with the according mechanism

I suppose that would also be OK... So you could have several possibilities:

1. The origin server doesn't send out the `interactionPolicy` key at all. `interactionPolicy.[canLike|canReply|canAnnounce|canQuote].always` are all implicitly understood to be their default values, as documented above. By doing this the origin server indicates it doesn't care (maybe doesn't even know) about interaction policies, and certainly not about verification.
2. The origin server sets only the keys it cares about, for example `interactionPolicy.canQuote.*`, while leaving out others like `interactionPolicy.[canLike|canReply|canAnnounce]`. This indicates that server only cares (maybe only knows about) verification of the activities for the field(s) it's set (`canQuote` in this case). All the unset ones are implicitly understood to be their default values, as documented above.
3. The origin server sets all keys `interactionPolicy.can*.*`. This indicates that the server knows about and cares about all of the listed ones, and verification thereof. So basically the same as 2, but without relying on any default values assumptions.

This would mean that the server has a way of indicating specific policy types it cares about verification for, while also indicating that it doesn't care about other types and probably won't process or send out any verification for them.

I think that covers `presence of the policy implying that it should be verified with the according mechanism`. What do you think?

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Author: Btw I was a bit confused in our earlier discussion because I thought you'd prefer it if the presence of an `interactionPolicy` (or specific keys thereof) *didn't* say anything about what sort of verification the server expects. I actually prefer it if it *does*, as you suggested. So I think maybe we ended up arriving at something we're both happier with, in a roundabout way?

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Reviewer: Basically, I think omitting a key should imply the status quo, even if you include other keys.

E.g. setting `interactionPolicy.canQuote.always` while leaving out `interactionPolicy.canReply` says nothing about replies, and thus, in the existing world where replies are established as not needing any kind of reply mechanism, would mean replies are allowed, and no `Accept` will be issued for them.

Which is distinct from setting `interactionPolic.canReply.always: as:Public`, because that would mean that you still use the underlying approval mechanism.

I think we do agree on that after all?

That being said, this also means there is little reason to bundle everything in an `interactionPolicy` hash, as you will have to check each policy individually anyway.

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Author: > I think omitting a key should imply the status quo, even if you include other keys. [...] I think we do agree on that after all?

Got you :) And yes we do!! I'll change the document.

> there is little reason to bundle everything in an interactionPolicy hash

Well, it's just a bit tidier to nest it isn't it? And it connotes that `interactionPolicy` is a coherent Thing. It's like with the `endpoints` object on Actors, which contains `sharedInbox` and other stuff. You could just put `sharedInbox` at the top-level json, but it makes sense to group it with the other endpoints as part of an `endpoints` object. Same principle here, imo.

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Author: I've updated the PR!

I split the interaction policy document into a separate markdown file as it was becoming LARGE.

You can see the proposal rendered nicely here: https://gotosocial--3703.org.readthedocs.build/en/3703/federation/interaction_policy/

In particular, I added this section about how to indicate that a server knows about and does / expects verification of policies (enforces them, in other words): https://gotosocial--3703.org.readthedocs.build/en/3703/federation/interaction_policy/#indicating-that-verification-is-required-not-required-per-sub-policy

I also introduced the term "sub-policy" to denote `canLike`, `canReply`, `canQuote` etc. Taken together, these sub-policies comprise an interaction policy.

I'm pretty happy with this version, and I think it covers all the points raised... Could you take a final pass through it and let me know what you think? Still v. happy to make changes if you see anything :)

> That is, the presence of the `interactionPolicy` property on a post **does not** necessarily imply that the server from which the post originated will send out an `Accept` to remote interacters. This is to allow for gradual adoption of `interactionPolicy` by server softwares in order to codify and make explicit **existing implicit assumptions** about permitted interactions, without putting a burden on implementers to add the `Accept`/`Reject` dance to their software immediately.

Author: Oh right this bit is also important: https://gotosocial--3703.org.readthedocs.build/en/3703/federation/interaction_policy/#set-target-property-on-accept-and-reject

}
```

In the `interactionPolicy` object:

- `canLike` indicates who is permitted to create a `Like` with the post URI as the `Object` of the `Like`.
- `canReply` indicates who is permitted to create a post with `inReplyTo` set to the URI of the post.
- `canAnnounce` indicates who is permitted to create an `Announce` with the post URI as the `Object` of the `Announce`. 

And:

> - `always` is an array of ActivityPub URIs/IDs of `Actor`s or `Collection`s of `Actor`s who are permitted to create + distribute an interaction without necessarily having to wait for an `Accept` from the interactee's server (more on this below).

Author: Right, I see what you mean now :) I'm not sure how to word this better, do you have any suggestions?

> - `always` is an array of ActivityPub URIs/IDs of `Actor`s or `Collection`s of `Actor`s who are permitted to create + distribute an interaction without necessarily having to wait for an `Accept` from the interactee's server (more on this below).

Reviewer: Maybe the following?
```suggestion
- `always` is an array of ActivityPub URIs/IDs of `Actor`s or `Collection`s of `Actor`s who are permitted to create + distribute an interaction without manual approval from the interacted-with user.
```

The fact that it's not a verification mechanism is why I'm reluctant with it being distinct from `approvalRequired` since at the protocol level everything would require approval. But I also understand this may be useful UX-wise.

> - `always` is an array of ActivityPub URIs/IDs of `Actor`s or `Collection`s of `Actor`s who are permitted to create + distribute an interaction without necessarily having to wait for an `Accept` from the interactee's server (more on this below).

Author: Like it! Committed it to the branch :)

!!! tip
  Note that **the below list is not exhaustive**, and extensions to `interactionPolicy` may wish to define **different defaults** for other types of interaction. For example, the default for `canQuote` may limit quoting only to the author themself if the property `interactionPolicy.canQuote` is not set.

##### `canLike`

If `canLike` is missing on an `interactionPolicy`, or the value of `canLike` is `null` or `{}`, then implementations should assume `"canLike": { "always": [ "https://www.w3.org/ns/activitystreams#Public" ] }`, ie., **anyone who can see the post can like it**.

##### `canReply`

If `canReply` is missing on an `interactionPolicy`, or the value of `canReply` is `null` or `{}`, then implementations should assume `"canReply": { "always": [ "https://www.w3.org/ns/activitystreams#Public" ] }`, ie., **anyone who can see the post can reply to it**.

##### `canAnnounce`

If `canAnnounce` is missing on an `interactionPolicy`, or the value of `canAnnounce` is `null` or `{}`, then implementations should assume `"canAnnounce": { "always": [ "https://www.w3.org/ns/activitystreams#Public" ] }`, ie., **anyone who can see the post can announce it**.

!!! tip

>     As is standard across AP implementations, implementations may still wish to limit `Announce` actities targeting the post to only the author themself if the post is addressed to followers-only.

Author: Fixed!

### Specifying Nobody

!!! note
    GoToSocial makes implicit assumptions about who can/can't interact, even if a policy specifies nobody. See [implicit assumptions](#implicit-assumptions).

To specify that **nobody** can perform an interaction on a post **except** for its author (who is always permitted), implementations should set the `always` array to **just the URI of the post author**, and `approvalRequired` can be unset, `null`, or empty.

For example, the following `canLike` value indicates that nobody can `Like` the post it is attached to except for the post author:

```json
"canLike": {

>   "always": [ "the_activitypub_uri_of_the_post_author" ],

Reviewer: The ActivityStreams spec says (https://www.w3.org/TR/activitystreams-core/#jsonld):

> The serialized JSON form of an Activity Streams 2.0 document MUST be consistent with what would be produced by the standard JSON-LD 1.0 Processing Algorithms and API [[JSON-LD-API](https://www.w3.org/TR/activitystreams-core/#bib-JSON-LD-API)] Compaction Algorithm using, at least, the normative JSON-LD `@context` definition

Depending on how you interpret it, it might mean you can't have a single-element array. But that's nitpicky.

Author: Oh yeah true. At https://gotosocial.org/ns these properties are defined as being `@id` which indeed can be just strings or objects or arrays right? Maybe it would be better to say that they should be a collection (so that they're always serialized as arrays); is that what you mean?

Reviewer: I think it is fine either way, but if keeping the current definition, I would just have the single string in the example so that implementers have a heads up that it may be rendered like that.

Author: Fixed it :)
